### ISSUE TITLE
Unable to get word level time stamps from azure speech to text api

### ISSUE THREAD

**Issue by SCLIIITB** [Azure-Samples/cognitive-services-speech-sdk#2156]
**IN ORDER TO ASSIST YOU, PLEASE PROVIDE THE FOLLOWING:**
Unable to get the word level time stamps. I tried setting the specchConfig.requestWordLevelTimestamps() but its giving an error saying that the method does not exist in speech Config. I also tried setting the property "speechConfig.setProperty("SpeechServiceResponse_RequestWordLevelTimestamps","true")" but this is not working either. In the output json, "Words" array is coming as null.
- Speech SDK log taken from a run that exhibits the reported issue. 
The method requestWordLevelTimestamps() is undefined for the type SpeechConfig

- A stripped down, simplified version of your source code that exhibits the issue. Or, preferably, try to reproduce the problem with one of the public samples in this repository (or a minimally modified version of it), and share the code.

package com.iml.gyan.speech.litassist.tts;

import java.util.concurrent.Future;

import javax.json.Json;
import javax.json.JsonArray;
import javax.json.JsonObject;
import javax.json.JsonReader;

import com.microsoft.cognitiveservices.speech.SpeechConfig;
import com.microsoft.cognitiveservices.speech.SpeechRecognitionResult;
import com.microsoft.cognitiveservices.speech.SpeechRecognizer;
import com.microsoft.cognitiveservices.speech.audio.AudioConfig;
import com.microsoft.cognitiveservices.speech.*;
import com.microsoft.cognitiveservices.speech.audio.*;

import java.io.StringReader;
import java.util.Scanner;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

public class azureSTT {
    // This example requires environment variables named "SPEECH_KEY" and "SPEECH_REGION"
	private static String speechSubscriptionKey = "";
    private static String serviceRegion = "";

    public static String getSTT(String audioFile, String lang) {
    	String res = null;
        try {
            //String audioFile = "<path to wav file>/SentenceTest.wav";
        	AudioConfig audioConfig = AudioConfig.fromWavFileInput(audioFile);
            SpeechConfig speechConfig = SpeechConfig.fromSubscription(speechSubscriptionKey, serviceRegion);

            speechConfig.setProperty("format", "detailed"); //you have to do it in this order
            speechConfig.setOutputFormat(OutputFormat.Detailed);
//            speechConfig.setProperty("SpeechServiceResponse_RequestWordLevelTimestamps","true");
            speechConfig.requestWordLevelTimestamps();
            speechConfig.setSpeechRecognitionLanguage(lang);//"en-US");
            SpeechRecognizer speechRecognizer = new SpeechRecognizer(speechConfig, audioConfig);
            System.out.println("######:"+speechConfig.getProperty("SpeechServiceResponse_RequestWordLevelTimestamps"));
            
            Future<SpeechRecognitionResult> task = speechRecognizer.recognizeOnceAsync();
            SpeechRecognitionResult result = task.get();
            
            // Now get the detailed recognition results as a JSON string
            String jsonText = result.getProperties().getProperty(PropertyId.SpeechServiceResponse_JsonResult);
            
            // Convert the JSON string to a JSON object
            JsonReader jsonReader = Json.createReader(new StringReader(jsonText));

            // Extract the "NBest" array of recognition results from the JSON.
            // Note that the first cell in the NBest array corresponds to the recognition results
            // (NOT the cell with the highest confidence number!)
            JsonArray nbestArray = jsonReader.readObject().getJsonArray("NBest");

            for (int i = 0; i < nbestArray.size(); i++) {
                JsonObject nbestItem = nbestArray.getJsonObject(i);
                System.out.println("\tConfidence: " + nbestItem.getJsonNumber("Confidence"));
                System.out.println("\tLexical: " + nbestItem.getString("Lexical"));
                // ITN stands for Inverse Text Normalization
                System.out.println("\tITN: " + nbestItem.getString("ITN"));
                System.out.println("\tMaskedITN: " + nbestItem.getString("MaskedITN"));
                System.out.println("\tDisplay: " + nbestItem.getString("Display"));
                System.out.println("\tWords Size: " + nbestItem.getJsonArray("Words"));
                // Word-level timing
                JsonArray wordsArray = nbestItem.getJsonArray("Words");
                System.out.println("\tWords Size: " + wordsArray.size());
                System.out.println("\t\tWord | Offset | Duration");
                for (int j = 0; j < wordsArray.size(); j++) {
                    JsonObject wordItem = wordsArray.getJsonObject(j);
                    System.out.println("\t\t" + wordItem.getString("Word") + " " + wordItem.getJsonNumber("Offset") + " " + wordItem.getJsonNumber("Duration"));
                }
            }

            jsonReader.close();
            res = result.getText();
//            System.out.println("RECOGNIZED: Text=" + );
//            recognizer.close();
        } catch (Exception ex) {
            System.out.println("Azure STT not working: " + ex.getMessage());
        }
		return res;
    }
    public static void main(String[] args) {
        String audioFile = "C:/Users/IML-WKS-Admin/Documents/Dec16/literacy-assistant-engine-api/literacy-assistant-engine-apis/resource/SentenceTest.wav";

    	System.out.println(getSTT(audioFile, "en-US"));
    }

}

- If relevant, a WAV file of your input audio.

- Additional information as shown below

**Describe the bug**

Unable to get the word level time stamps. I tried setting the speechConfig.requestWordLevelTimestamps() but its giving an error saying that the method does not exist in speech Config. I also tried setting the property "speechConfig.setProperty("SpeechServiceResponse_RequestWordLevelTimestamps","true")" but this is not working either. In the output json, "Words" array is coming as null.

**To Reproduce**

Run the code given above.

**Expected behavior**

A json file having recognised text and also word level timestamps.

**Version of the Cognitive Services Speech SDK**

1.33.0

**Platform, Operating System, and Programming Language**

 - OS:  Windows
 - Hardware - x64
 - Programming language: Java, Maven project
 - Browser: NA

**Additional context**

 - output of the speech to text:
 ######:
	Confidence: 0.88268495
	Lexical: friday and romeo was so excited
	ITN: friday and romeo was so excited
	MaskedITN: friday and romeo was so excited
	Display: Friday and Romeo was so excited.
	Words Size: null
Azure STT not working: Cannot invoke "javax.json.JsonArray.size()" because "wordsArray" is null
null

**Comment by jhakulin** [comment]
@SCLIIITB Could you please try with the https://github.com/Azure-Samples/cognitive-services-speech-sdk/blob/master/samples/java/jre/console/src/com/microsoft/cognitiveservices/speech/samples/console/SpeechRecognitionSamples.java sample and there recognitionWithMicrophoneAsyncAndDetailedRecognitionResults sample and let us know if you can get the word level timestamp information using that correctly?

**Comment by SCLIIITB** [comment]
Hi @jhakulin I tried running the same code and getting an error on line 97 in the code given on the above link you provided. The error is : The method requestWordLevelTimestamps() is undefined for the type SpeechConfig. I imported SpeechConfig from : import com.microsoft.cognitiveservices.speech.SpeechConfig;


Please help me in this!

Thank You

**Comment by jhakulin** [comment]
@SCLIIITB Hmm, sounds strange, need to investigate a bit more about the error you are seeing and if I am able to reproduce it. However, I do not think you need to call config.requestWordLevelTimestamps(); to get the word level information in JSON result if you already set config.setOutputFormat(OutputFormat.Detailed) which is done in that sample also and that should be enough to get the detailed results in JSON including word level timestamps. Please try removing the call and if that does not help, could you please share the exact error you are getting ? You should have been able to call the method, so I wonder if the real reason is some kind of installation problem of Speech SDK maven package.

**Comment by jhakulin** [comment]
@SCLIIITB did you had success with the previous proposal?

**Comment by jhakulin** [comment]
@SCLIIITB The problem seems to be that for certain input the SpeechServiceResponse_JsonResult may contain nBest fields which are empty result and do not have words array and this is not handled correctly in the sample. Below example input where this might happen:

Say something...
RECOGNIZED: Text = Hello. Hello.
Offset: 10700000
Duration: 7200000
{"Id":"c27289029ef7428f875485dd7f856926","RecognitionStatus":"Success","Offset":10700000,"Duration":7200000,"Channel":0,"DisplayText":"Hello. Hello.","NBest":[{"Confidence":0.61016554,"Lexical":"hello hello","ITN":"hello hello","MaskedITN":"hello hello","Display":"Hello. Hello.","Words":[{"Word":"hello","Offset":10700000,"Duration":2800000,"Confidence":0.46928927},{"Word":"hello","Offset":13500000,"Duration":4400000,"Confidence":0.7510418}]},{"Confidence":0.0,"Lexical":"","ITN":"","MaskedITN":"","Display":""},{"Confidence":0.43132108,"Lexical":"hello","ITN":"hello","MaskedITN":"hello","Display":"hello","Words":[{"Word":"hello","Offset":10700000,"Duration":7200000,"Confidence":0.43132108}]}]}
        Confidence: 0.61016554
        Lexical: hello hello
        ITN: hello hello
        MaskedITN: hello hello
        Display: Hello. Hello.
                Word | Offset | Duration
                hello 10700000 2800000
                hello 13500000 4400000
        Confidence: 0.0
        Lexical:
        ITN:
        MaskedITN:
        Display:
                Word | Offset | Duration
Unexpected java.lang.NullPointerException: Cannot invoke "jakarta.json.JsonArray.size()" because "wordsArray" is null

There we had "{"Confidence":0.0,"Lexical":"","ITN":"","MaskedITN":"","Display":""}" fields under nBest array that do not have words as service returned extra empty result. We need to handle this in sample code better and do not try to go through the words in case words array is null. See below code which you can apply. We will improve the sample in the meantime. I believe service can occasionally send empty result after the recognized results.

```
                for (int i = 0; i < nbestArray.size(); i++) {
                    JsonObject nbestItem = nbestArray.getJsonObject(i);
                    System.out.println("\tConfidence: " + nbestItem.getJsonNumber("Confidence"));
                    System.out.println("\tLexical: " + nbestItem.getString("Lexical"));
                    // ITN stands for Inverse Text Normalization
                    System.out.println("\tITN: " + nbestItem.getString("ITN"));
                    System.out.println("\tMaskedITN: " + nbestItem.getString("MaskedITN"));
                    System.out.println("\tDisplay: " + nbestItem.getString("Display"));

                    // Check if "Words" key exists and is not null
                    JsonArray wordsArray = nbestItem.getJsonArray("Words");
                    if (wordsArray != null) {
                        System.out.println("\t\tWord | Offset | Duration");
                        for (int j = 0; j < wordsArray.size(); j++) {
                            JsonObject wordItem = wordsArray.getJsonObject(j);
                            System.out.println("\t\t" + wordItem.getString("Word") + " " + wordItem.getJsonNumber("Offset") + " " + wordItem.getJsonNumber("Duration"));
                        }
                    } else {
                        System.out.println("\t\tNo word-level timing information available.");
                    }
                }
```

**Comment by jhakulin** [comment]
Closing the issue, workaround provided, sample will be fixed in 1.34.0 release beginning of Dec.

**Comment by SCLIIITB** [comment]
Hi @jhakulin I tried multiple times with different audio files but I coudn't get the words array. Words array is always null. I check my installation as well and its fine. I hope we don't have to set any parameter on the azure portal to enable the word level information. Below is the code I am trying:

`package com.iml.gyan.speech.litassist.tts;

import java.util.concurrent.Future;

import javax.json.Json;
import javax.json.JsonArray;
import javax.json.JsonObject;
import javax.json.JsonReader;

import com.microsoft.cognitiveservices.speech.SpeechConfig;
import com.microsoft.cognitiveservices.speech.SpeechRecognitionResult;
import com.microsoft.cognitiveservices.speech.SpeechRecognizer;
import com.microsoft.cognitiveservices.speech.audio.AudioConfig;
import com.microsoft.cognitiveservices.speech.*;
import com.microsoft.cognitiveservices.speech.audio.*;

import java.io.StringReader;
import java.util.Scanner;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

public class azureSTT {
    // This example requires environment variables named "SPEECH_KEY" and "SPEECH_REGION"
	private static String speechSubscriptionKey = "";
    private static String serviceRegion = "eastus";

    public static String getSTT(String audioFile, String lang) {
    	String res = null;
        try {
          
            AudioConfig audioConfig = AudioConfig.fromWavFileInput(audioFile);
            SpeechConfig speechConfig = SpeechConfig.fromSubscription(speechSubscriptionKey, serviceRegion);

            speechConfig.setProperty("format", "detailed"); //you have to do it in this order
            speechConfig.setOutputFormat(OutputFormat.Detailed);
            speechConfig.setSpeechRecognitionLanguage(lang);//"en-US");
            SpeechRecognizer speechRecognizer = new SpeechRecognizer(speechConfig, audioConfig);
            
            Future<SpeechRecognitionResult> task = speechRecognizer.recognizeOnceAsync();
            SpeechRecognitionResult result = task.get();
            
            // Now get the detailed recognition results as a JSON string
            String jsonText = result.getProperties().getProperty(PropertyId.SpeechServiceResponse_JsonResult);
            
            // Convert the JSON string to a JSON object
            JsonReader jsonReader = Json.createReader(new StringReader(jsonText));

            JsonArray nbestArray = jsonReader.readObject().getJsonArray("NBest");

            for (int i = 0; i < nbestArray.size(); i++) {
                JsonObject nbestItem = nbestArray.getJsonObject(i);
                System.out.println("\tConfidence: " + nbestItem.getJsonNumber("Confidence"));
                System.out.println("\tLexical: " + nbestItem.getString("Lexical"));
                // ITN stands for Inverse Text Normalization
                System.out.println("\tITN: " + nbestItem.getString("ITN"));
                System.out.println("\tMaskedITN: " + nbestItem.getString("MaskedITN"));
                System.out.println("\tDisplay: " + nbestItem.getString("Display"));
                System.out.println("\tWords Size: " + nbestItem.getJsonArray("Words"));
                // Word-level timing
                JsonArray wordsArray = nbestItem.getJsonArray("Words");
                System.out.println("\tWords Size: " + wordsArray.size());
                System.out.println("\t\tWord | Offset | Duration");
                for (int j = 0; j < wordsArray.size(); j++) {
                    JsonObject wordItem = wordsArray.getJsonObject(j);
                    System.out.println("\t\t" + wordItem.getString("Word") + " " + wordItem.getJsonNumber("Offset") + " " + wordItem.getJsonNumber("Duration"));
                }
            }

            jsonReader.close();
            res = result.getText();
//            System.out.println("RECOGNIZED: Text=" + );
//            recognizer.close();
        } catch (Exception ex) {
            System.out.println("Azure STT not working: " + ex.getMessage());
        }
		return res;
    }
    public static void main(String[] args) {
        String audioFile = <path to audio file //default_2892238173549351111199121111111111111_0__180.wav";

    	System.out.println(getSTT(audioFile, "en-US"));
    }

}
`

**Below is the output which i get:**

	Confidence: 0.9877272
	Lexical: bob is a frog who lives in my yard
	ITN: bob is a frog who lives in my yard
	MaskedITN: bob is a frog who lives in my yard
	Display: Bob is a frog who lives in my yard.
	Words Size: null
Azure STT not working: Cannot invoke "javax.json.JsonArray.size()" because "wordsArray" is null
null
